src/julia_to_gap.jl

```diff
@@ -93,148 +131,213 @@ function julia_to_gap(x::Rational{T}) where {T<:Integer}
             return -Globals.infinity
         end
     end
-    numer = julia_to_gap(numer_julia)
-    denom = julia_to_gap(denom_julia)
+    numer = GapObj_internal(numer_julia, cache, Val(recursive))
+    denom = GapObj_internal(denom_julia, cache, Val(recursive))
     return Wrappers.QUO(numer, denom)
 end
 
 ## Floats
-julia_to_gap(x::Float64) = NEW_MACFLOAT(x)
-julia_to_gap(x::Float32) = NEW_MACFLOAT(Float64(x))
-julia_to_gap(x::Float16) = NEW_MACFLOAT(Float64(x))
+GAP.@install GapObj(x::Float64) = NEW_MACFLOAT(x)
+GAP.@install GapObj(x::Float32) = NEW_MACFLOAT(Float64(x))
+GAP.@install GapObj(x::Float16) = NEW_MACFLOAT(Float64(x))
 
 ## Chars
-julia_to_gap(x::Char) = CharWithValue(Cuchar(x))
+GAP.@install GapObj(x::Char) = CharWithValue(Cuchar(x))
 
 ## Strings and symbols
-julia_to_gap(x::AbstractString) = MakeString(string(x))
-julia_to_gap(x::Symbol) = MakeString(string(x))
+GAP.@install GapObj(x::AbstractString) = MakeString(string(x))
+GAP.@install GapObj(x::Symbol) = MakeString(string(x))
+
+# helper functions for recursion
+function recursion_info(::Type{T}, obj, recursive::Bool, recursion_dict::GapCacheDict) where {T}
+    recursive && recursion_dict !== nothing && haskey(recursion_dict, obj) && return true, false, recursion_dict
+    rec = recursive && _needs_tracking_julia_to_gap(T)
+    if rec && recursion_dict === nothing
+        rec_dict = RecDict()
+    else
+        rec_dict = recursion_dict
+    end
+
+    return false, rec, rec_dict
+end
 
-## Generic caller for optional arguments
-julia_to_gap(obj::Any; recursive::Bool) = julia_to_gap(obj)
-julia_to_gap(obj::Any, recursion_dict::IdDict{Any,Any}; recursive::Bool = true) = julia_to_gap(obj)
+function handle_recursion(obj, ret_val, rec::Bool, rec_dict::GapCacheDict)
+    if rec_dict !== nothing
+        rec_dict[obj] = ret_val
+    end
+    return rec ? rec_dict : nothing
+end
 
 ## Arrays (including BitVector)
-function julia_to_gap(
+function GapObj_internal(
     obj::AbstractVector{T},
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
-) where {T}
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where {T, recursive}
+
+    ret, rec, rec_dict = recursion_info(T, obj, recursive, recursion_dict)
```

Comment:
"ret" and "rec" are very easy to mix up, they look so similar.

And then we also have `rec_dict` and `ret_val`...

I think we could use a macro to avoid introducing some of these. But all of this can be done in a follow-up PR (or not at all).